File: svc_Core/Bank.Core.App/Dto/AccountEventDto.cs
using Bank.Core.Domain.Events;

namespace Bank.Core.App.Dto;

public class AccountEventDto
{
    public Guid Id { get; set; }
    public DateTime ResolvedAt { get; protected set; }

    public long Value { get; set; }
    public string Comment { get; set; }

    public EventType Type { get; set; }

    public enum EventType
    {
        Transfer,
        Withdrawal,
        Deposit,
        CreditPayment
    }

    public AccountEventDto From(BalanceChange source)
    {
        return new AccountEventDto
        {
            Id = source.Id,
            ResolvedAt = source.ResolvedAt,
            Value = source.Value,
            Type = source switch
            {
                CreditPayment => EventType.CreditPayment,
                Withdrawal => EventType.Withdrawal,
                Deposit => EventType.Deposit,
                _ => throw new ArgumentOutOfRangeException(),
            },
            Comment = source.Comment,
        };
    }

    public AccountEventDto From(Transfer source)
    {
        return new AccountEventDto
        {
            Type = EventType.Transfer,
            Comment = source.Comment,
            Value = source.Value,
            ResolvedAt = source.ResolvedAt,
            Id = source.Id,
        };
    }
}

File: client_Customer/app/src/main/java/com/example/customerclient/data/api/dto/AccountEventDto.kt
package com.example.customerclient.data.api.dto

import com.example.customerclient.ui.bill.info.BillHistory
import com.example.customerclient.ui.bill.info.HistoryOperationType
import com.example.customerclient.ui.bill.info.OperationType

data class AccountEventDto(
    val id: String,
    val comment: String?,
    val eventType: String,
    val balanceChange: BalanceChangeDto?,
    val transfer: TransferDto?,
    val resolvedAt: String,
    val state: String
)

fun AccountEventDto.toBillHistory(): BillHistory {
    return when (this.eventType) {
        "Transfer" -> {
            BillHistory(
                id = this.id,
                type = OperationType.TRANSFER,
                eventType = when (this.transfer?.target?.eventType) {
                    "Withdrawal" -> HistoryOperationType.WITHDRAW
                    "Deposit" -> HistoryOperationType.TOP_UP
                    "0" -> HistoryOperationType.WITHDRAW
                    "1" -> HistoryOperationType.TOP_UP
                    else -> HistoryOperationType.WITHDRAW
                },
                amount = "${this.transfer?.target?.nativeValue?.amount} ${this.transfer?.target?.nativeValue?.currency}",
                date = this.resolvedAt.split("T")[0],
                billId = this.balanceChange?.accountId ?: ""
            )
        }

        "BalanceChange" -> {
            BillHistory(
                id = this.id,
                type = OperationType.BALANCE_CHANGE,
                eventType = when (this.balanceChange?.eventType) {
                    "Withdrawal" -> HistoryOperationType.WITHDRAW
                    "Deposit" -> HistoryOperationType.TOP_UP
                    "0" -> HistoryOperationType.WITHDRAW
                    "1" -> HistoryOperationType.TOP_UP
                    else -> HistoryOperationType.WITHDRAW
                },
                amount = "${this.balanceChange?.nativeValue?.amount} ${this.balanceChange?.nativeValue?.currency}",
                date = this.resolvedAt.split("T")[0],
                billId = this.balanceChange?.accountId ?: ""
            )
        }

        else -> {
            BillHistory(
                id = this.id,
                eventType = when (this.balanceChange?.eventType) {
                    "Withdrawal" -> HistoryOperationType.WITHDRAW
                    "Deposit" -> HistoryOperationType.TOP_UP
                    "0" -> HistoryOperationType.WITHDRAW
                    "1" -> HistoryOperationType.TOP_UP
                    else -> HistoryOperationType.WITHDRAW
                },
                amount = "${this.balanceChange?.nativeValue?.amount} ${this.balanceChange?.nativeValue?.currency}",
                date = this.resolvedAt.split("T")[0],
                billId = this.balanceChange?.accountId ?: ""
            )
        }
    }
}
